### PR TITLE
Low precision layer norm option

### PR DESCRIPTION
**Pull request by dakinggg**
Verified that the model print out has the right layer for true, false and default. Verified that the first few steps of training proceed without crashing, and loss goes down.

### CODE REVIEW

File: examples/llm/src/models/layers/attention.py [hunk 1]
```diff
@@ -140,8 +142,10 @@ def __init__(self, cfg: DictConfig, device: Optional[str] = None):
             self.out_proj._is_residual = True  # type: ignore
 
             if self.attn_qk_ln:
-                self.q_ln = nn.LayerNorm(self.d_model, device=device)
-                self.k_ln = nn.LayerNorm(self.d_model, device=device)
+                self.layernorm_class = torch.nn.LayerNorm if not cfg.get(
```
Comment (vchiley) [inline examples/llm/src/models/layers/attention.py:145]:
```suggestion
                self.layernorm_class = nn.LayerNorm if not cfg.get(
```

we're imported nn

File: examples/llm/src/models/layers/attention.py [hunk 2]
```diff
@@ -247,8 +251,10 @@ def __init__(self, cfg: DictConfig, device: Optional[str] = None):
...
```
Comment (vchiley) [inline examples/llm/src/models/layers/attention.py:254]:
```suggestion
                self.layernorm_class = nn.LayerNorm if not cfg.get(
```

we're imported nn

File: examples/llm/src/models/layers/gpt_blocks.py [hunk 3]
```diff
@@ -36,9 +38,13 @@ def __init__(self,
         super().__init__()
         if cfg.get('alibi', False):
             assert cfg.attn_impl == 'triton' or cfg.attn_impl == 'torch', 'Only triton kernel or torch supports alibi'
-        self.ln_1 = nn.LayerNorm(cfg.d_model, device=device)
+
+        self.layernorm_class = torch.nn.LayerNorm if not cfg.get(
```
Comment (vchiley) [inline examples/llm/src/models/layers/gpt_blocks.py:42]:
```suggestion
        self.layernorm_class = nn.LayerNorm if not cfg.get(
```

we're imported nn

File: examples/llm/src/models/mosaic_gpt.py [hunk 4]
```diff
@@ -37,6 +39,9 @@ def __init__(self, cfg: DictConfig):
         else:
             raise ValueError(f'Unknown attn_impl={cfg.attn_impl}')
 
+        self.layernorm_class = torch.nn.LayerNorm if not cfg.get(
```
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:42]:
```suggestion
        self.layernorm_class = nn.LayerNorm if not cfg.get(
```

we're imported nn

**Review by vchiley — APPROVED** [review]
LGTM

Some nit comments
also why set `self.layernorm_class`? does it need to be attached to the obj instance?

**Comment by dakinggg** [comment]
It does not need to be attached, except that it is used multiple times and I thought cleaner this way, but not a strong opinion

File: examples/llm/src/models/layers/attention.py (hunk 1 above)
Comment (dakinggg) [reply examples/llm/src/models/layers/attention.py:145]: oops, thanks

**Comment by dakinggg** [comment]
unattached it
